doc/SONiC_Port_Mirroring_HLD.md

```diff
@@ -0,0 +1,319 @@
+  * [7. Warm Boot Support](#7-warm-boot-support)
+  * [8. Scalability](#8-scalability)
+  * [9. Unit Test](#9-unit-test)
+
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+
+# Revision
+| Rev |     Date    |       Author       | Change Description                         |
+|:---:|:-----------:|:------------------:|--------------------------------------------|
+| 0.1 | 05/17/2019  |   Rupesh Kumar      | Initial version                            |
+
+
+# About this Manual
+This document provides general information about extending mirroring implementation in SONiC.
+# Scope
+This document describes the high level design of Mirroring Enhancements feature. 
+
+
+# Definition/Abbreviation
+### Table 1: Abbreviations
+| **Term**                 | **Meaning**                         |
+|--------------------------|-------------------------------------|
+|   SPAN                   |  Switched Port ANalyzer                |
+|   ERSPAN                 |  Encapsulated Remote Switched Port ANalyzer                |
+
+
+# 1 Requirement Overview
+## 1.1 Functional Requirements
+
+1. Port/Port-channel mirroring support
+     - Add support to mirror ingress traffic on port/port-channel to SPAN/ERPSAN mirror session.
+     - Add support to mirror egress traffic on port/port-channel to SPAN/ERSPAN mirror session.
+     - Add support to mirror both ingress/egress traffic on port/port-channel to SPAN/ERSPAN mirror session.
+
+2. Dynamic session management
+    - Allow multiple source to single destination.
+    - Each session supports mirroring from single port to single destination port.
+    - Session-id created in SAI per destination port will be used when the same destination port is configured in other session.
+      This effectively utilizes the hardware resource to be shared across multiple sessions.
+    - Mirror session on source portchannel will be active if at least one port is part of portchannel.
+    - Mirror session on source portchannel will become inactive when portchannel has no members.
+    - ERSPAN session will be active/inactive based on destination IP reachability.
+
+3. ACL rules can continue to use port/ERSPAN sessions as the action.
+
+4. Configuration CLI for mirror session
+    - CLI allows all flavors of mirror sessions.
+    - CLI validation for all mandatory parameters in ERSPAN configuration.
+    - CLI validation for all mandatory parameters in port/portchannel mirroring.
+    - CLI to allow mirror session configuration only with destination port.
+
+
+## 1.2 Configuration and Management Requirements
+- Existing CLI 'config mirror_session add/remove'to be extended to include source port/portchannel.
+- Existing CLI 'config mirror_session add/remove' to be extended to include destination port/portchannel.
+- Existing CLI 'show mirror session' is extended to support all flavors of mirror sessions.
+
+
+## 1.3 Scalability Requirements
+- Up to max ASIC capable mirror sessions to be supported.
```

Comment:
how do we know how many mirror session can be supported in the asic?

Reply:
The SAI Attribute SAI_SWITCH_ATTR_MAX_MIRROR_SESSION can be used to support the max mirror sessinons. This is for both ERSPAN and SPAN and we don't have any mechanism in orchagent to retrieve this now.

Reply:
Do we have a number based on TD3 and TH series?
Also how many sessions can be active at the same time. Is there any limitation

Reply:
For TD3 and TH series, max 4 sessions can be active at same time. single session can be shared across multiple source ports.